src/loaders/jobs/taskfarmer/taskfarmer_task_mgr.py

```diff
@@ -0,0 +1,357 @@
+                elif parsed_status == 'CG':
+                    return JobStatus.CANCELLED
+                else:
+                    raise ValueError(f"Unrecognized job status: {squeue_out}")
+            else:
+                return JobStatus.FAILED
+
+    def _append_task_info(self, task_info):
+        """
+        Append the task information to the task info file
+        """
+
+        if not all(key in task_info for key in REQUIRED_TASK_INFO_KEYS):
+            raise ValueError(f"task_info must contain all keys: {REQUIRED_TASK_INFO_KEYS}")
+
+        task_info.update({'kbase_collection': self.kbase_collection, 'load_ver': self.load_ver, 'tool': self.tool,
+                          'source_data_dir': self.source_data_dir})
+
+        # Append the new record to the file
+        task_info_file = self._get_task_info_file()
+        with open(task_info_file, 'a') as writer:
+            fcntl.flock(writer.fileno(), fcntl.LOCK_EX)
+            json.dump(task_info, writer)
+            writer.write('\n')
+            fcntl.flock(writer.fileno(), fcntl.LOCK_UN)
+
+    def _task_exists(self):
+        """
+        Check if the task for kbase_collection, load_ver, tool has been submitted.
+        """
+
+        if self.force_run:
+            # The job directory is recreated as part of the initialization process.
+            job_dir_exists = True
+        else:
+            # check if the job directory exists and is not empty
+            job_dir_exists = os.path.isdir(self.job_dir) and os.listdir(self.job_dir)
+
+        # check task exists in the task info file
+        task_info_file = self._get_task_info_file()
+        df = pd.read_json(task_info_file, lines=True)
+
+        if df.empty:
+            return False
+
+        tasks_df = df[(df["kbase_collection"] == self.kbase_collection) &
+                      (df["load_ver"] == self.load_ver) &
+                      (df["tool"] == self.tool)]
+
+        return job_dir_exists and not tasks_df.empty
+
+    def _cancel_job(self, job_id):
+        """
+        Cancel a job on NERSC.
+        """
+
+        job_status, retry = self._get_job_status_from_nersc(job_id), 0
+
+        if job_status in [JobStatus.COMPLETED, JobStatus.FAILED, JobStatus.CANCELLED]:
+            return True
+
+        while job_status in [JobStatus.RUNNING, JobStatus.PENDING] and retry < 3:
+            retry += 1
+            print(f"Canceling job {job_id} (try #{retry})...")
+            tf_common.run_nersc_command(
+                ['scancel', str(job_id)], self.job_dir, log_file_prefix='scancel', check_return_code=True)
+            time.sleep(2)
+            # Get the job status to confirm the job is canceled
+            job_status = self._get_job_status_from_nersc(job_id)
+
+        if job_status in [JobStatus.RUNNING, JobStatus.PENDING]:
+            raise ValueError(f"Failed to cancel job {job_id}.")
+        else:
+            print(f"Job {job_id} is canceled.")
+
+    def _get_latest_task(self):
```

Comment:
I think you could merge this method into `_retrieve_latest_task_status` as the latter is always called directly after this method. The return could be the task info with the status added into a new dict key

Reply:
👍

Reply:
:+1: